refactor: Request Language Resolution

fahimalizain opened the pull request:
### Priority order for resolving `frappe.local.lang`

1. Form Dict => _lang
2. Cookie => preferred_language
3. Request Header => Accept-Language
4. User document => language
5. System Settings => language

### Changes

* Refactored HTTPRequest class under `frappe.auth`
* Updated the priority of language settings
* Clear `preferred_language` key from cookie in `post_login`

---

Please provide your thoughts, this is an issue in v12 too.

fahimalizain committed:
fix: frappe.local.lang resolution

gavindsouza committed:
perf: Skip guess_language if _lang is provided

gavindsouza reviewed (commented):
The order of setting the language for the session I observed was

1. User language: Language set in User document 
2. System language: System Settings
3. Cookie: preferred_language
4. Request Header: Accept-Language

I tested this on develop and your PR, couldn't observe any change. Checked `curl` output and used `print` statements to confirm the value of `frappe.local.lang` set. 

Any pointers for testing this out?

fahimalizain commented:
@gavindsouza thank you for the review.
I have been making requests by setting `Accept-Language` header. This is my observation on latest develop (88cb19c)

I placed a `print(frappe.local.lang)` right after `init_request()` in `frappe.app`.
I have been invoking `cmd: ping`

For Guest users:
- Sending in a `ping` request with `Accept-Language: ru` prints `ru` ✅

For Authenticated Users - Users with a session
- Sending in a `ping` request with `Accept-Language: ru` prints the language `user.lang || system_settings.lang` ❌
- `Accept-Language` header and `preferred_language` cookie seems to have zero effect on the lang of an Authenticated User.

Regarding the order of setting the language for the session you observed, which one has the highest priority ? User.lang or Header.AcceptLang ?

Thank you

fahimalizain commented:
ping!

gavindsouza commented:
Hey @fahimalizain, from my observation the current order is as the numbered list I've mentioned in my previous comment. User.lang has the highest priority and the header has the lowest.

But this doesn't seem right. I think the order should be reversed. Also, this needs more discussion. 

cc: @frappe/frappe-team

gavindsouza commented:
> The order of setting the language for the session I observed was
> 
> 1. User language: Language set in User document
> 2. System language: System Settings
> 3. Cookie: preferred_language
> 4. Request Header: Accept-Language

@fahimalizain Discussed with the internal team, we agree that the priority should be 4 -> 3 -> 1 -> 2 instead of the current order.

fahimalizain committed:
fix: preferred_language cookie support for all users

fahimalizain commented:
@gavindsouza please check now.
`preferred_language` cookie is applicable to authenticated users too now, and validated the order of lang resolution.

gavindsouza committed:
refactor: Simplify HTTPRequest class

* For the sake of Readability and ease of understanding
* Style updates

gavindsouza committed:
refactor: Set Language in HTTPHeader

Order of priority for setting language:
1. Form Dict => _lang
2. Cookie => preferred_language
3. Request Header => Accept-Language
4. User document => language
5. System Settings => language

Cookie is placed at #2 since the language picker in the navbar depends
on it. And the Accept-Language header sends values based on the client's locales.

---

Form Dict _lang now accepts language codes too. Previously, language
names were used...for whatever reason.

gavindsouza committed:
Merge branch 'develop' into lang-header-cookie

gavindsouza committed:
refactor: Rename guess_language as get_language

Guess suggests there's some AI involvement. The get_language function
has a defined priority. It is deterministic, hence teh name change.

gavindsouza committed:
fix: Don't redefine datetime

* Sort imports
* Update file header

gavindsouza committed:
fix: Clear preferred_language cookie post login

If preferred_language was set in cookie pre login, clear it after a
successful login so that User or Site specific settings can be applied

gavindsouza commented:
@fahimalizain Sorry about highjacking your PR. I might need to accomodate more changes to make this merge-able 😅 . 

I may raise another PR with all the changes in this branch instead of continuing this way. Thank you for bringing attention to this issue.

gavindsouza commented:
Continuing via https://github.com/frappe/frappe/pull/13703
